pkg/tools/etcd_helper.go

```diff
@@ -469,7 +471,16 @@ func (h *EtcdHelper) SetObj(key string, obj, out runtime.Object, ttl uint64) err
 
 // Pass an EtcdUpdateFunc to EtcdHelper.GuaranteedUpdate to make an etcd update that is guaranteed to succeed.
 // See the comment for GuaranteedUpdate for more detail.
-type EtcdUpdateFunc func(input runtime.Object) (output runtime.Object, ttl uint64, err error)
+type EtcdUpdateFunc func(input runtime.Object, node *etcd.Node, res *etcd.Response) (output runtime.Object, ttl *uint64, err error)
```

Comment:
Why would we need a generic abstraction around etcd when we already have generic.etcd?  If I were writing a DB backend for this I wouldn't reuse guaranteed update, but would probably need to provide tools that bridge those concepts from my underlying schema.  Helper still seems like a bridge between higher level object semantics and lower level kv semantics.  Things like TTL and the raft index are really not generic concepts (certainly not the total ordering assumptions)

> On May 23, 2015, at 1:32 AM, Daniel Smith <notifications@github.com> wrote:
> 
> In pkg/tools/etcd_helper.go:
> 
> > @@ -469,7 +471,16 @@ func (h *EtcdHelper) SetObj(key string, obj, out runtime.Object, ttl uint64) err
> >  
> >  // Pass an EtcdUpdateFunc to EtcdHelper.GuaranteedUpdate to make an etcd update that is guaranteed to succeed.
> >  // See the comment for GuaranteedUpdate for more detail.
> > -type EtcdUpdateFunc func(input runtime.Object) (output runtime.Object, ttl uint64, err error)
> > +type EtcdUpdateFunc func(input runtime.Object, node *etcd.Node, res *etcd.Response) (output runtime.Object, ttl *uint64, err error)
> I'd strongly prefer not to expose etcd.Node outside of this package, because I'd like to evolve this into a generic storage interface. Can we provide only object & ttl as input?
> 
> (arguably the proper way to do this is to encode ttl into our objects, in metadata or as an annotation-- then none of these interfaces would need to change, but that's much more work of course)
> 
> —
> Reply to this email directly or view it on GitHub.
>